docs/secrets.md

```diff
@@ -0,0 +1,65 @@
+
+**Fine Grained Authorization**:  keys will generally not be accessible in the flux language.  A secret-enabled function will accept meta-data about the secret (a key) and use it to look up the stored value in memory. The lookup will error if the key is not found or if the user is not authorized to use that key.  
+
+# **Design**
+
+## Internal Functions
+A function that requires secret lookup must: 
+
+- Be built-in to the language
+- Register an ExternalDependencies struct  (note: it will be straightforward and advantageous to refactor the `InjectFromDependencies` and similar functions into this pattern as well).   
+- Not return secrets under any circumstances (e.g. in error messages) 
+
+The external interface for secret lookup is: 
+```Go
+func LookupRequest(keyObject values.Object) SecretRequest
+func LookupKey(req SecretRequest) (string, error)
+```
+The keyObject is a values.Object type because: 
+
+- The secret parameter can be passed to the service without being modified
+- The parameter may be validated by the service, minimizing the dependency of the calling function to have prior knowledge of the key lookup protocol. 
+
+The keyObject should minimally have a `secretKey` property: 
+slackID = {secretKey: "FLUX_SLACK_ID"}
+
+but the object may include supplementary information for consumption by the calling function.  They may also be embedded + replaced in another values.Value type (e.g. an encapsulating Object or Array) by the calling function as in:
+
+http.post(header: {"Content Type": "application-json", "ApplicationSecret": {secretKey: "FLUX_HTTP_POST_KEY"}}, url: "http://some/url/", data: json_enc)
+
+## Flux Package
+We will provide a `vault` package that will help the user create an object with the required properties for secret lookup.  
+A helper function is useful because we may wish to change the properties of a secret request over time, and we don't want users to have to update their code unless strictly needed.  
+
+The vault package contains: 
+
+- `vault.get(key: "FLUX_SLACK_ID")`: returns an object with all properties needed to make a request, e.g.,  `{userID: "11918", orgID: "19119384", secretKey: "my-secret"}`
+
+The function will provide all that's needed to do the lookup.  It may also be implemented to pre-authorize that the user can access the given key, so that errors occur earlier.  
+
+Finally, a nice-to-have, but not required behavior would be to enforce that the token-lookup object be created by the vault.Get call.  That is, the user should get an error if they construct a request object by hand.  
+
+A final implementation may look like: 
+```Go
+// calling flux:  
+//     import "vault"
+//     import "slack"
+//
+//     tok = vault.get(key: "FLUX_SLACK_ID")
```

Comment:
If I follow what you're saying, `tok` is just an ordinary object. It has properties and they can be accessed just like any other object. And it doesn't contain any privileged information. Functions that consume this object however must be implemented in Go because they will use the information in that object to actually obtain the privileged information and use it.

Is this correct?

Reply:
that's correct